### PR TITLE
feat(query): add date_diff function

### PR DESCRIPTION
**Pull request by chagelo**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

related to #16534

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/16560)

### CODE REVIEW

**Comment by chagelo** [comment]
very confusing, I have registered the date_diff's different func name like diff_years, but it gives me `no function matches signature add_years(Date, Date)` error, passed the unittest, but cann't pass the sqllogic test.

File: src/query/functions/tests/it/scalars/testdata/datetime.txt [hunk 1]
```diff
@@ -3345,3 +3345,291 @@ output domain  : {1630812366000000..=1630812366000000}
...
+output         : 1
+
+
+ast            : date_diff(year, to_timestamp('2020-02-29 23:59:59.165575'), to_timestamp('2019-02-28 23:59:59.423179'))
+raw expr       : diff_years(to_timestamp('2019-02-28 23:59:59.423179'), to_timestamp('2020-02-29 23:59:59.165575'))
+checked expr   : diff_years<Timestamp, Timestamp>(to_timestamp<String>("2019-02-28 23:59:59.423179"), to_timestamp<String>("2020-02-29 23:59:59.165575"))
+optimized expr : 0_i64
+output type    : Int64
+output domain  : {0..=0}
+output         : 0
```
Comment (TCeason) [inline src/query/functions/tests/it/scalars/testdata/datetime.txt:3390]: select date_diff(year, to_timestamp('2020-02-29 23:59:59.165575'), to_timestamp('2019-02-28 23:59:59.423179'))

expect -1

File: src/query/functions/tests/it/scalars/testdata/datetime.txt [hunk 2]
```diff
@@ -3345,3 +3345,291 @@ output domain  : {1630812366000000..=1630812366000000}
...
+output         : 5
+
+
+ast            : date_diff(quarter, to_timestamp('2020-02-29 23:59:59.165575'), to_timestamp('2019-02-28 23:59:59.423179'))
+raw expr       : diff_quarters(to_timestamp('2019-02-28 23:59:59.423179'), to_timestamp('2020-02-29 23:59:59.165575'))
+checked expr   : diff_quarters<Timestamp, Timestamp>(to_timestamp<String>("2019-02-28 23:59:59.423179"), to_timestamp<String>("2020-02-29 23:59:59.165575"))
+optimized expr : -3_i64
+output type    : Int64
+output domain  : {-3..=-3}
+output         : -3
```
Comment (TCeason) [inline src/query/functions/tests/it/scalars/testdata/datetime.txt:3435]: I test in snowflake , `select datediff(quarter, to_timestamp('2020-02-29 23:59:59.165575'), to_timestamp('2019-02-28 23:59:59.423179'))` expect -4

File: src/query/functions/tests/it/scalars/testdata/datetime.txt [hunk 3]
```diff
@@ -3345,3 +3345,291 @@ output domain  : {1630812366000000..=1630812366000000}
...
+output         : 24
+
+
+ast            : date_diff(year, to_timestamp('2023-11-12 09:38:18.165575'), to_timestamp('2025-03-27 21:01:35.423179'))
+raw expr       : diff_years(to_timestamp('2025-03-27 21:01:35.423179'), to_timestamp('2023-11-12 09:38:18.165575'))
+checked expr   : diff_years<Timestamp, Timestamp>(to_timestamp<String>("2025-03-27 21:01:35.423179"), to_timestamp<String>("2023-11-12 09:38:18.165575"))
+optimized expr : 1_i64
+output type    : Int64
+output domain  : {1..=1}
+output         : 1
```
Comment (TCeason) [inline src/query/functions/tests/it/scalars/testdata/datetime.txt:3381]: select datediff(year, to_timestamp('2023-11-12 09:38:18.165575'), to_timestamp('2025-03-27 21:01:35.423179')) expect 2.

File: tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test [hunk 4]
```diff
@@ -901,6 +901,41 @@ select date_add(QUARTER, 1, to_date('2018-01-02'))
...
+
+query T
+SELECT DATE_DIFF(YEAR, to_date('2000-01-02'), to_date('2024-01-01'))
+----
+23
+
+query T
+SELECT DATE_DIFF(YEAR, to_date('2012-02-29'), to_date('2013-02-28'))
+----
+1
```
Comment (TCeason) [inline tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test:917]: consider test select DATE_DIFF(YEAR, TO_TIMESTAMP('2023-12-31 09:38:18.165575'), TO_TIMESTAMP('2024-01-01 21:01:35.423179')) the result should be 1.

Comment (TCeason) [reply tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test:917]: Also need to add some tests where the sign is negative

**Comment by chagelo** [comment]
Sorry, I got you. But, what's the result of `select date_diff(year, to_date('2023-12-31'), to_date('2024-01-01'))`?

**Comment by TCeason** [comment]
> Sorry, I got you. But, what's the result of `select date_diff(year, to_date('2023-12-31'), to_date('2024-01-01'))`?

Oh, I just mean add a logic test case:

```
query T
SELECT DATE_DIFF(YEAR, to_date('2023-12-31'), to_date('2024-01-01'))
----
1

```

**Comment by chagelo** [comment]
So what should I do is to fix it to be compatible with Snowflake and add more tests?

**Comment by TCeason** [comment]
> So what should I do is to fix it to be compatible with Snowflake and add more tests?

Date diff about year just need to evaluate year sub. E.g. 

2023-12-31 2024-01-01 just need to calc year sub.

**Comment by chagelo** [comment]
Ok, thank you.

TCeason ***@***.***> 于2024年10月8日周二 14:33写道：

File: src/query/expression/src/utils/date_helper.rs [hunk 5]
```diff
@@ -362,35 +365,157 @@ macro_rules! impl_interval_year_month {
...
+        date_end.year() as i64 - date_start.year() as i64
+    }
+}
+
+pub struct EvalQuartersImpl;
+
+impl EvalQuartersImpl {
+    pub fn eval_date_diff(date_start: i32, date_end: i32, tz: TzLUT) -> i32 {
+        let diff_months = EvalMonthsImpl::eval_date_diff(date_start, date_end, tz);
+        diff_months.signum() * diff_months.abs().div_ceil(3)
```
Comment (TCeason) [inline src/query/expression/src/utils/date_helper.rs:390]: Now Databen has ToQuarter calc quartrter.

```rust
impl ToNumber<u8> for ToQuarter {
    fn to_number(dt: &DateTime<Tz>) -> u8 {
        (dt.month0() / 3 + 1) as u8
    }
}
```

https://github.com/databendlabs/databend/blob/a58d8b42aec4d38c2cffd40d2e6eb9ac2cf6bc90/src/query/expression/src/utils/date_helper.rs#L499

Directly calc diff and div 3 maybe cause some mistake.

**Review by TCeason — APPROVED** [review]

File: tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test [hunk 6]
```diff
@@ -901,6 +901,176 @@ select date_add(QUARTER, 1, to_date('2018-01-02'))
...
+----
+0
+
+query T
+SELECT DATE_DIFF(QUARTER, to_date('2023-06-01'), to_date('2024-01-01'))
+----
+3
+
+query T
+SELECT DATE_DIFF(QUARTER, to_date('2024-01-01'), to_date('2023-06-01'))
```
Comment (TCeason) [inline tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test:940]: Good case. 

```

// Special case handling: If the end date is in the first quarter of the next year, and the start date is in the second quarter or later of the current year, then subtract 1.


let adjusted_quarters2 = if end_date2.year() > start_date2.year() && end_date2.month() <= 3 && start_date2.month() >= 4 {
quarters2 - 1
} else {
quarters2
};

```

**Review by TCeason — CHANGES_REQUESTED** [review]

**Review by TCeason — APPROVED** [review]
LGTM

**Comment by TCeason** [comment]
Hi @chagelo  after approving this pr, I see you force-push a commit. What's diff in this commit? just rebase lastest main?

image

**Comment by chagelo** [comment]
Yes

**Comment by BohuTANG** [comment]
@soyeric128 Documentation need update. Thanks.

**Comment by BohuTANG** [comment]
@chagelo Thanks to your contribution!
